pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/AvgShedder.java

```diff
@@ -0,0 +1,320 @@
+                        localBrokerData.getMsgThroughputIn() + localBrokerData.getMsgThroughputOut(),
+                        localBrokerData.getMsgRateIn() + localBrokerData.getMsgRateOut());
+            }
+        }
+
+        // sort brokers by scores.
+        brokers.sort((e1, e2) -> (int) (brokerScoreMap.get(e1) - brokerScoreMap.get(e2)));
+    }
+
+    private Double calculateScores(LocalBrokerData localBrokerData, final ServiceConfiguration conf) {
+        return localBrokerData.getMaxResourceUsageWithWeight(
+                conf.getLoadBalancerCPUResourceWeight(),
+                conf.getLoadBalancerDirectMemoryResourceWeight(),
+                conf.getLoadBalancerBandwidthInResourceWeight(),
+                conf.getLoadBalancerBandwidthOutResourceWeight()) * 100;
+    }
+
+    private List<Pair<String, String>> findBrokerPairs(List<String> brokers,
+                                                       double lowThreshold, double highThreshold) {
+        List<Pair<String, String>> pairs = new LinkedList<>();
+        int i = 0, j = brokers.size() - 1;
+        while (i <= j) {
+            String maxBroker = brokers.get(j);
+            String minBroker = brokers.get(i);
+            if (brokerScoreMap.get(maxBroker) - brokerScoreMap.get(minBroker) < lowThreshold) {
+                brokerHitCountForHigh.remove(maxBroker);
+                brokerHitCountForHigh.remove(minBroker);
+
+                brokerHitCountForLow.remove(maxBroker);
+                brokerHitCountForLow.remove(minBroker);
+            } else {
+                pairs.add(Pair.of(minBroker, maxBroker));
+                if (brokerScoreMap.get(maxBroker) - brokerScoreMap.get(minBroker) < highThreshold) {
+                    brokerHitCountForLow.computeIfAbsent(minBroker, k -> new MutableInt(0)).increment();
+                    brokerHitCountForLow.computeIfAbsent(maxBroker, k -> new MutableInt(0)).increment();
+
+                    brokerHitCountForHigh.remove(maxBroker);
+                    brokerHitCountForHigh.remove(minBroker);
+                } else {
+                    brokerHitCountForLow.computeIfAbsent(minBroker, k -> new MutableInt(0)).increment();
+                    brokerHitCountForLow.computeIfAbsent(maxBroker, k -> new MutableInt(0)).increment();
+
+                    brokerHitCountForHigh.computeIfAbsent(minBroker, k -> new MutableInt(0)).increment();
+                    brokerHitCountForHigh.computeIfAbsent(maxBroker, k -> new MutableInt(0)).increment();
+                }
+            }
+            i++;
+            j--;
+        }
+        return pairs;
+    }
+
+    @Override
+    public Optional<String> selectBroker(Set<String> candidates, BundleData bundleToAssign, LoadData loadData,
+                                         ServiceConfiguration conf) {
+        if (!bundleBrokerMap.containsKey(bundleToAssign) || !candidates.contains(bundleBrokerMap.get(bundleToAssign))) {
```

Comment:
```java
```
```suggestion
        final var brokerToUnload = bundleBrokerMap.getOrDefault(bundleToAssign, null);
        if (brokerToUnload == null || !candidates.contains(brokerToUnload)) {
```

Reduce a find operation on `bundleBrokerMap`, and then you can return `Optional.of(brokerToUnload)` in the `else` branch.

Reply:
Updated, thanks for review.